pykeops/torch/nystrom/nystrom.py

```diff
@@ -0,0 +1,265 @@
+                self.inv_eps = 1e-4
+            else:
+                self.inv_eps = 1e-8
+        if not mask_radius:
+            if kernel == "rbf":
+                self.mask_radius = 2 * np.sqrt(2) * self.sigma
+            if kernel == "exp":
+                self.mask_radius = 8 * self.exp_sigma
+
+    def fit(self, X: torch.tensor):
+        """
+        Args:   X = torch tensor with features of shape
+                (1, n_samples, n_features)
+        Returns: Fitted instance of the class
+        """
+
+        # Basic checks: we have a lazy tensor and n_components isn't too large
+        assert type(X) == torch.Tensor, "Input to fit(.) must be a Tensor."
+        assert (
+            X.size(0) >= self.n_components
+        ), "The application needs X.shape[1] >= n_components."
+        # self._update_dtype(X)
+        # Number of samples
+        n_samples = X.size(0)
+        # Define basis
+        rnd = check_random_state(self.random_state)
+        inds = rnd.permutation(n_samples)
+        basis_inds = inds[: self.n_components]
+        basis = X[basis_inds]
+        # Build smaller kernel
+        basis_kernel = self._pairwise_kernels(basis, kernel=self.kernel)
+        # Get SVD
+        U, S, V = torch.svd(basis_kernel)
+        S = torch.max(S, torch.ones(S.size()) * 1e-12)
+        self.normalization_ = torch.mm(U / np.sqrt(S), V.t())
+        self.components_ = basis
+        self.component_indices_ = inds
+
+        return self
+
+    def transform(self, X: torch.tensor) -> torch.tensor:
+        """Applies transform on the data.
+        Args:
+            X [LazyTensor] = data to transform
+        Returns
+            X [LazyTensor] = data after transformation
+        """
+        K_nq = self._pairwise_kernels(X, self.components_, self.kernel)
+        return K_nq @ self.normalization_.t()
+
+    def K_approx(self, X: torch.tensor) -> torch.tensor:
+        """Function to return Nystrom approximation to the kernel.
+        Args:
+            X[torch.tensor] = data used in fit(.) function.
+        Returns
+            K[torch.tensor] = Nystrom approximation to kernel"""
+
+        K_nq = self._pairwise_kernels(X, self.components_, self.kernel)
+        K_approx = K_nq @ self.normalization_ @ K_nq.t()
+        return K_approx
+
+    def _pairwise_kernels(
+        self, x: torch.tensor, y: torch.tensor = None, kernel="rbf", sigma: float = 1.0
+    ) -> LazyTensor:
+        """Helper function to build kernel
+        Args:   X = torch tensor of dimension 2.
+                K_type = type of Kernel to return
+        Returns:
+                K_ij[LazyTensor]
+        """
+        if y is None:
+            y = x
+        if kernel == "linear":
+            K_ij = x @ y.T
+        elif kernel == "rbf":
+            x /= sigma
+            y /= sigma
+
+            x_i, x_j = LazyTensor(x[:, None, :]), LazyTensor(y[None, :, :])
+            K_ij = (-1 * ((x_i - x_j) ** 2).sum(-1)).exp()
```

Comment:
Note that in KeOps, `- ((x_i - x_j) ** 2).sum(-1)` could be slightly more efficient and legible.

Reply:
done